// @dcl/schemas/dist/dapps/preview/preview-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMessage = exports.PreviewMessageType = void 0;
const validation_1 = require("../../validation");
var PreviewMessageType;
(function (PreviewMessageType) {
    PreviewMessageType["READY"] = "ready";
    PreviewMessageType["LOAD"] = "load";
    PreviewMessageType["ERROR"] = "error";
    PreviewMessageType["UPDATE"] = "update";
})(PreviewMessageType = exports.PreviewMessageType || (exports.PreviewMessageType = {}));
/** @alpha */
(function (PreviewMessageType) {
    PreviewMessageType.schema = {
        type: 'string',
        enum: Object.values(PreviewMessageType)
    };
    PreviewMessageType.validate = (0, validation_1.generateValidator)(PreviewMessageType.schema);
})(PreviewMessageType = exports.PreviewMessageType || (exports.PreviewMessageType = {}));
const sendMessage = (window, type, payload, targetOrigin = '*') => {
    const event = { type, payload };
    window.postMessage(event, targetOrigin);
};
exports.sendMessage = sendMessage;
