// decentraland-ecs/src/cli/mock-catalyst/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.mockCatalyst = void 0;
var path = require("path");
var http_proxy_middleware_1 = require("http-proxy-middleware");
var setupUtils_1 = require("../setupUtils");
var wearables_1 = require("../wearables");
var mockCatalyst = function (app, baseFolders) {
    serveFolders(app, baseFolders);
    app.get('/lambdas/explore/realms', function (req, res) {
        res.json([
            {
                serverName: 'localhost',
                url: "http://" + req.get('host'),
                layer: 'stub',
                usersCount: 0,
                maxUsers: 100,
                userParcels: []
            }
        ]);
    });
    app.get('/lambdas/contracts/servers', function (req, res) {
        res.json([
            {
                address: "http://" + req.get('host'),
                owner: '0x0000000000000000000000000000000000000000',
                id: '0x0000000000000000000000000000000000000000000000000000000000000000'
            }
        ]);
    });
    app.get('/lambdas/profiles', function (req, res, next) { return __awaiter(void 0, void 0, void 0, function () {
        var previewWearables, deployedProfile, err_1;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 5, , 6]);
                    return [4 /*yield*/, (0, wearables_1.getAllPreviewWearables)({
                            baseFolders: baseFolders,
                            baseUrl: ''
                        }).map(function (wearable) { return wearable.id; })];
                case 1:
                    previewWearables = _b.sent();
                    if (!(previewWearables.length === 1)) return [3 /*break*/, 4];
                    return [4 /*yield*/, fetch("https://peer.decentraland.org" + req.originalUrl)];
                case 2: return [4 /*yield*/, (_b.sent()).json()];
                case 3:
                    deployedProfile = _b.sent();
                    if ((deployedProfile === null || deployedProfile === void 0 ? void 0 : deployedProfile.length) === 1) {
                        (_a = deployedProfile[0].avatars[0].avatar.wearables).push.apply(_a, previewWearables);
                        return [2 /*return*/, res.json(deployedProfile)];
                    }
                    _b.label = 4;
                case 4: return [3 /*break*/, 6];
                case 5:
                    err_1 = _b.sent();
                    console.warn("Failed to catch profile and fill with preview wearables.", err_1);
                    return [3 /*break*/, 6];
                case 6: return [2 /*return*/, next()];
            }
        });
    }); });
    // fallback all lambdas to a real catalyst
    app.use('/lambdas', (0, http_proxy_middleware_1.createProxyMiddleware)({
        target: 'https://peer.decentraland.org/',
        changeOrigin: true,
        timeout: 25 * 1000,
        proxyTimeout: 25 * 1000,
        onError: function (err, req_, res) {
            console.warn("Oops, it seems the catalyst isn't working well.");
            res.writeHead(500);
            res.end('');
        }
    }));
    // fallback all lambdas to a real catalyst
    app.use('/content', (0, http_proxy_middleware_1.createProxyMiddleware)({
        target: 'https://peer.decentraland.org/',
        changeOrigin: true,
        timeout: 25 * 1000,
        proxyTimeout: 25 * 1000,
        onError: function (err, req_, res) {
            console.warn("Oops, it seems the content server isn't working well.");
            res.writeHead(500);
            res.end('');
        }
    }));
};
exports.mockCatalyst = mockCatalyst;
var serveFolders = function (app, baseFolders) {
    app.get('/content/contents/:hash', function (req, res, next) {
        if (req.params.hash && req.params.hash.startsWith('b64-')) {
            var fullPath_1 = path.resolve(Buffer.from(req.params.hash.replace(/^b64-/, ''), 'base64').toString('utf8'));
            // only return files IF the file is within a baseFolder
            if (!baseFolders.find(function (folder) { return fullPath_1.startsWith(folder); })) {
                next();
                return;
            }
            var options = {
                dotfiles: 'deny',
                maxAge: 1,
                cacheControl: false,
                lastModified: true,
                headers: {
                    'x-timestamp': Date.now(),
                    'x-sent': true,
                    etag: JSON.stringify(Date.now().toString()),
                    'cache-control': 'no-cache,private,max-age=1'
                }
            };
            res.sendFile(fullPath_1, options, function (err) {
                if (err) {
                    next(err);
                }
            });
        }
    });
    function pointerRequestHandler(pointers) {
        if (!pointers) {
            return [];
        }
        var requestedPointers = new Set(pointers && typeof pointers === 'string'
            ? [pointers]
            : pointers);
        var resultEntities = (0, setupUtils_1.getSceneJson)({
            baseFolders: baseFolders,
            pointers: Array.from(requestedPointers)
        });
        return resultEntities;
    }
    app.get('/content/entities/scene', function (req, res) {
        return res.json(pointerRequestHandler(req.query.pointer)).end();
    });
    app.post('/content/entities/active', function (req, res) {
        return res.json(pointerRequestHandler(req.body.pointers)).end();
    });
};
